manifests/init.pp

```diff
@@ -149,18 +149,30 @@
   # TODO: doesn't allow deploying a Pulp non-mirror without Foreman
   $shared_with_foreman_vhost = !$pulpcore_mirror
 
+  if $shared_with_foreman_vhost {
+    include foreman::config::apache
+    Class['foreman::config::apache'] -> Class['pulpcore::apache']
+    $servername = $foreman::config::apache::servername
+  } else {
+    include certs::apache
+    Class['certs::apache'] ~> Class['pulpcore::apache']
+    $servername = $certs::apache::hostname
+  }
+
   $rhsm_port = $reverse_proxy_real ? {
     true  => $reverse_proxy_port,
     false => 443
   }
 
+  $rhsm_url = "https://${servername}:${rhsm_port}/rhsm"
```

Comment:
What is the advantage of constructing this here, and then deconstructing it again here (https://github.com/theforeman/puppet-foreman_proxy_content/pull/384/files#diff-0393165ecfe8896424ac06417b31d37ab529a477ea8982c4573945a31a28e43aR20-R22)? 

I am looking to understand the benefit of adding the complexity of deconstruction when we were already passing the individual parameters to the class. That seems to add code into the process that could be avoided by passing the parts in directly as we already do today.

I would add on to that, that we plan to drop the boostrap RPM in a few coming releases, so thats a factor in just how much support code we add in this area.

Reply:
My thinking has mostly been around treating `$rhsm_url` as a single parameter and eventually single source of truth for the `rhsm_hostname`, `rhsm_port`, and `rhsm_path`.

I considered (and previously wrote it as) changing the interface to the bootstrap RPM type to accept the `$rhsm_url` and use `URI::Parse` inside the script template to perform the deconstruction there. I explored this approach instead to avoid the breaking change to that type interface, moving the deconstruction outside of the provider.

But the way I currently wrote it does instead cause a breaking change in the `foreman_proxy_content::bootstrap_rpm` class so I agree it's kind of pointless. I am tempted to preserve that class interface as well and perform the deconstruction inside `init.pp` in order to preserve interfaces while ensuring that there is a single source of truth for `$rhsm_url` and its constituent pieces that will be propagated to the bootstrap RPM and smart_proxy_pulp plugin.

Longer term I had the idea of exposing some parameters in the `foreman_proxy_content` class, that may default to `undef` but if defined would take precedence over `$servername`, `$reverse_proxy_real ? { true => $reverse_proxy_port, false => 443 }`, and `/rhsm` (perhaps there is little use for that last one though) in the way the `$rhsm_url` gets constructed. I see this as being more forward thinking towards the possibility of HA services and a more modular design for the SmartProxy. (It would also require separating the rhsm_url out of the smartproxy_pulp plugin and into @ekohl 's idea for smartproxy_rhsm plugin)

Reply:
I also want to state that I have some skepticism about the removal of the bootstrap RPM, based on separation in some deployments between katello users (meaning someone who know what katello is, has a login) vs. their end users (maybe someone in a different team or department, or sometimes even an external customer) who may be a sysadmin that doesn't know about or have any access to touch Katello at all, and only knows that in order to get subscriptions and repositories they do: `rpm -Uvh http://smartproxy.example.com/pub/katello-ca-consumer-latest.noarch.rpm && subscription-manager register --activationkey=$key1,$key2,$key3`

For that reason, my preference would be not to remove the functionality of the bootstrap RPM entirely. I understand that global registration is a better process for a heavily automated, cloud first world, but many users, and particularly many of the _user's users_, are not there yet and might prefer not to deprecate the current registration workflow. I think I recall that there are some other concerns with the bootstrap RPM as well. Perhaps keeping it around in an opt-in fashion (i.e. `Boolean generate_and_deploy_bootstrap_rpm = false` in the interface for the main `foreman_proxy_content` class) is still better to serve a wider swath of users?

Reply:
I think if there are concerns over moving to global registration as the one and only method for how we do this procedure we should discuss that more broadly. Looking through our RFCs I do not see one dedicated to the discussion of deprecating these various workflows. I will look at opening one and then you can elevate your thoughts there for broader discussion.

Reply:
Here is said RFC -- https://community.theforeman.org/t/rfc-deprecating-katello-bootstrap-client-tools/25220